### PR TITLE
CEXT 1487

### PR DESCRIPTION
**Pull request by ShehzadAhm**
Implement auto-importing files for create and update commands 

## Description
Local development epic:
Right now, the files need to be copied, stringified, and minified before we add them to the meshConfig. This can get cumbersome.

To solve this problem, as part of this ticket, we are extending the CLI to auto-import the files by parsing the mesh config before creating or updating a mesh.

**Steps followed:**

1. Parse the mesh config and make a list of files that need to be imported. 
    - Identify where the files may be present in our MeshConfig
2. Validate file paths and throw an error with a list of all files that are wrong.
   - Check valid file types: js, json
   - Check if the length of file names is less than 25
   - If the files exist in the files array, check if it matches with the fileName in sources, transforms, additionalResolvers, or else error out.
3. If the file content is already present in the mesh config, take confirmation from the user if they want to override the content of the existing file. If "Yes", override else stop the operation.
4. If paths defined in source, transform or additionalResolvers are valid and they don’t exist in the files array, import those files, stringify, and minify them before adding them to the files array.

## Related Issue
[CEXT-1487](https://jira.corp.adobe.com/browse/CEXT-1487)

## Motivation and Context
To solve the problem of manually copying, stringifying, and minifying the files before we add them to the meshConfig, and hence, helping the customer to have CLI auto-import the files during create/update

## How Has This Been Tested?

The above feature has been tested using aio cli commands:

## Screenshots (if appropriate):
a. Create command returns error if file names are long (>25 characters)
image

b. Create command returns error if the files are not of JS, JSON type
image

c. Create command returns error if the files are not in the same directory as meshConfig (for not fully qualified meshConfig)
image

d. Create command returns error if the file names are not matching in meshConfig (in the below example '_schemaBody.json_' and '_schemaBodyTest.json_'
image

e. Create command imports the files to meshConfig (the files are appended in below example)
image
image

f. Create command prompts the user if they want to override the files in meshConfig and proceeds to prompt the user to create the mesh or not
image
image
Note that, in the above example, the schemaBody.json file is appended to meshConfig.

g. For fully-qualified meshConfig, if the files do not exist in file system, it should prompt to create the mesh
image
image

h. Create command fails if the file is outside the mesh directory (eg. '../additional-resolvers.js')
image

i. Create command fails if the file path starts from home directory (in Linux - '~' operator)
image

The above scenarios are also supported by the update command.

 
## Types of changes
1. Extended the create/update commands to support auto import
2. Extended helper classes to define new helper functions for validation & auto-import
3. Updated the unit tests for create & update

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by revanth0212** [comment]
I am yet to review the code but please make sure to merge these changes to `epic/local-dev` branch instead.

File: src/utils.js [hunk 1]
```diff
@@ -53,9 +58,222 @@ const jsonFlag = Flags.boolean({
...
+			};
+		} else {
+			//if the files array does not exist
+			if (!data.meshConfig.files) {
+				data.meshConfig.files = [];
+			}
+
+			//if the files arrray exists, we append the file path and content in meshConfig
+			data.meshConfig.files.push({
+				path: `./${path.basename(file)}`,
```
Comment (dthampy) [inline src/utils.js:255]: What happens if the file is in a different location than where the meshConfig is , like a parent directory?
I don’t think we should hardcode the path to the current directory where meshConfig isPlease make sure that if the file path is available, use that for setting files.path 
For example, 
```
"additionalResolvers": [
  "../additional-resolvers-2.js"
]
```

File: src/utils.js [hunk 2]
```diff
@@ -53,9 +58,222 @@ const jsonFlag = Flags.boolean({
...
+		throw new Error(
+			`Mesh file names must be less than 25 characters. The following file(s) are invalid: ${filesWithInvalidNames}.`,
+		);
+	}
+
+	// check if the the filePaths in the files array match
+	// the fileNames in sources, transforms or additionalResolvers
+	if (data.meshConfig.files) {
+		for (let i = 0; i < data.meshConfig.files.length; i++) {
+			if (filesList.indexOf(data.meshConfig.files[i].path) == -1) {
```
Comment (dthampy) [inline src/utils.js:210]: Shouldn't we be using strict equality operator? check for other places also

**Comment by revanth0212** [comment]
Please make sure to not throw any errors if the user provided a fully qualified mesh config but the files are not provided in the file system. It is not necessary to have the files on file system if they are already provided in the mesh config.

image

File: src/commands/api-mesh/create.js [hunk 3]
```diff
@@ -69,6 +74,27 @@ class CreateCommand extends Command {
 			);
 		}
 
+		let filesList = [];
+
+		try {
+			filesList = getFilesInMeshConfig(data, args.file);
+		} catch (err) {
+			this.log(err.message);
+			this.error('Input mesh config is not valid.');
```
Comment (dthampy) [inline src/commands/api-mesh/create.js:83]: When does this error get thrown? is it when the file is not found in the location specified? if so, please throw the error like "File not found in the location specified" which would help the user better.

File: src/utils.js [hunk 4]
```diff
@@ -53,9 +58,222 @@ const jsonFlag = Flags.boolean({
...
+ */
+function validateFilePaths(filesList, meshConfigName) {
+	try {
+		for (let file of filesList) {
+			// throw error, if the meshConfig and the file are not in the same directory
+			if (!fs.existsSync(path.resolve(path.dirname(meshConfigName), file))) {
+				throw new Error(
+					`Please make sure the files ${path.basename(file)} and ${path.basename(
+						meshConfigName,
+					)} are in the same directory.`,
```
Comment (dthampy) [inline src/utils.js:148]: I believe we can give the flexibility to the user to create the source file in the project root directory, not just in the directory where they have stored the meshConfig json file

**Comment by dthampy** [comment]
@Shehzad1399 , Good work! I have left few comments

**Comment by revanth0212** [comment]
I havent reviewed the code yet but I believe the experience can be adjusted. If you have any questions, we can catch up over a call.

1. Looks like the CLI is throwing an error even if the user has a fully qualified mesh. CLI should only try to import if the mesh has links but missing its contents in the files array
image
2. Looks like you are modifying the mesh config file if the user does not have file contents in the mesh. That is not necessary. We dont need to update the mesh config. Please change that.
image

**Comment by ShehzadAhm** [comment]
> I havent reviewed the code yet but I believe the experience can be adjusted. If you have any questions, we can catch up over a call.
> 
> 1. Looks like the CLI is throwing an error even if the user has a fully qualified mesh. CLI should only try to import if the mesh has links but missing its contents in the files array
> 
> image
> 
> 2. Looks like you are modifying the mesh config file if the user does not have file contents in the mesh. That is not necessary. We dont need to update the mesh config. Please change that.
> image

1. Made the changes to support creation of fully qualified meshConfigif the files are not on the fileSystem.
2. Updated the code to not update the meshConfig

File: src/commands/api-mesh/create.js (hunk 3 above)
Comment (ShehzadAhm) [reply src/commands/api-mesh/create.js:83]: This error gets thrown during below scenarios:
1. if the files are not of js, json type
image

2. if the file names are too long
image

File: src/utils.js (hunk 1 above)
Comment (ShehzadAhm) [reply src/utils.js:228]: If it is not a fully-qualified meshConfig, we are restricting the files to the same folder as mesh config.
To be checked with Revanth.

Comment (ShehzadAhm) [reply src/utils.js:228]: Done

**Comment by ShehzadAhm** [comment]
> Please make sure to not throw any errors if the user provided a fully qualified mesh config but the files are not provided in the file system. It is not necessary to have the files on file system if they are already provided in the mesh config.
> 
> image

Fixed

**Review by dthampy — APPROVED** [review]

**Review by revanth0212 — APPROVED** [review]

**Comment by revanth0212** [comment]
Nice work @ShehzadAhm 

Please resolve conflicting files before merging

**Pull request merged**